src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,828 @@
+//
+// The shorter string (corresponding to options), may have only a
+// few distinct values, so we need to mix it up with the longer
+// string in various ways.
+
+std::string even_bytes(const std::string& a) {
+  std::stringstream ss;
+  size_t len = a.length();
+  for (size_t j = 0; j < len; j += 2) {
+    ss << a[j];
+  }
+  return ss.str();
+}
+
+std::string hash_two_strings(const std::string& a, const std::string& b) {
+  bool a_shorter = (a.length() < b.length());
+  const std::string* shorter = (a_shorter ? &a : &b);
+  const std::string* longer  = (a_shorter ? &b : &a);
+  uint64_t hashA = std::hash<std::string>()(*longer);
+  uint64_t hashB = std::hash<std::string>()(*shorter + even_bytes(*longer));
+  if (sizeof(size_t) < 8) {
+    // On a 32bit machine we need to work harder to get 80 bits
+    uint64_t hashC = std::hash<std::string>()(*longer + "SmallChangeForThirdHash");
+    hashA += (hashC << 32);
+  }
+  if (a_shorter) hashA ^= 0xff; // order of strings should change result
+  hashA ^= ALL_HEADER_CKSUM; // checksum from all header files
+  hashA ^= (0x3ac5*hashB); // mix low bits of hashB into hashA
+  hashB &= 0xffff;
+  char buf[128];
+  char* out = buf;
+  for (int pos = 80; (pos -= 4) >= 0;) {
+    int64_t nibble = ((pos >= 64) ? (hashB >> (pos-64)) : (hashA >> pos)) & 0xf;
+    *out++ = ((nibble < 10) ? nibble+'0' : nibble-10+'a');
+  }
+  *out = 0;
+  return std::string(buf);
+}
+
+bool file_stat(const std::string& name, struct stat* st) {
+  // Open file for reading to avoid inconsistent cached attributes
+  int fd = ::open(name.c_str(), O_RDONLY, 0666);
+  if (fd < 0) return false;
+  int rc;
+  do {
+    errno = 0;
+    rc = ::fstat(fd, st);
+  } while ((rc < 0) && (errno == EINTR));
+  ::close(fd);
+  return (rc == 0);
+}
+
+bool file_exists_and_is_recent(const std::string& name) {
+  time_t now = ::time(nullptr);
+  struct stat st;
+  return (file_stat(name, &st) && (st.st_ctime+kBuildTimeoutSecs > now));
+}
+
+bool file_exists(const std::string& name) {
+  struct stat st;
+  return file_stat(name, &st);
+}
+
+std::string read_file_as_string(const std::string& name) {
+  FILE* f = fopen(name.c_str(), "r");
+  if (!f) return std::string("");
+  std::stringstream ss;
+  for (;;) {
+    int c = fgetc(f);
+    if (c == EOF) break;
+    ss << (char)c;
+  }
+  fclose(f);
+  return ss.str();
+}
+
+std::string getenv_with_default(const char* name, const char* dval) {
+  const char* s = ::getenv(name);
+  return std::string(s ? s : dval);
+}
+
+std::string run_shell_get_first_line(const char* cmd) {
+  FILE* f = popen(cmd, "r");
+  char buf[1024];
+  size_t len = 0;
+  if (f) {
+    for (;;) {
+      int c = fgetc(f);
+      if (c == EOF) break;
+      if ((c != '\n') && (len+1 < sizeof(buf))) buf[len++] = c;
+    }
+    pclose(f);
+  }
+  buf[len] = 0;
+  return std::string(buf);
+}
+
+std::string get_java_home() {
+  auto s = run_shell_get_first_line(
+    "java -XshowSettings:properties -version 2>&1 | fgrep -i java.home"
```

Comment:
Put this logic in the Makefile, it's way simpler there.  The generated Makefile and the libhail build Makefile should share common logic.

Reply:
Put it in the Makefile.  The only thing I want you to shell out to is make.

In general, for a first PR, we want the simplest, robust thing that works and that other people can develop on.  We can always change it later, and I expect it to.  And when the process is working, it's quick: a simpler PR-takes a few minutes and the users have it 40m later (which I still think is too slow).

I want this simpler, simpler:
 - rip out the cache (I'll comment above),
 - rip out the multi-threading, it won't help:
 - the master is single-threaded so the ModuleBuilder doesn't even need to lock,
 - there should be a single lock when entering from Java for the workers.  They are all running the same code and need to load the same modules in the same order, so that's going to happen serially anyway.
 - Also, I've said many times, I want to avoid multi-threaded code if at all possible and I'm going to push hard on designs that require it.

Reply:
Strongly disagree but you're driving.  II'll give it a single global mutex.

See comments above re cache and the relationship of NativeModule to DLL files.

Reply:
> Strongly disagree

Acknowledged.